### PR TITLE
[skinning] Skin timer implementation

### PR DESCRIPTION
**Pull request by enen92**
## Motivation and context
Currently time-based logic in the skinning engine is pretty limited: skins either rely on the `System.IdleTime` boolean condition (which is linked to input) for checking "idle" state which doesn't necessarily mean no actions on that time, dummy windows ([ref](https://github.com/b-jesch/skin.estuary.modv2/blob/Nexus/xml/Custom_1110_CloseVideoOSD.xml)) for auto closing other windows, trigger the `AlarmClock` builtin with skin properties (see [ref](https://github.com/xbmc/xbmc/pull/20816)) or simply delegate time-based stuff to background python scripts. All of these solutions are quite hacky and some (such as python) have a performance penalty that may not be acceptable for a default skin or low-end hardware.

I've created this with 3 main use cases in mind:

- [x] Remove the ugly AlarmClock/Property base implementation of https://github.com/xbmc/xbmc/pull/20816 (**edit:** Done in https://github.com/xbmc/xbmc/pull/21328)
- [x] Implement the autoclose video OSD setting in Estuary (alternative to https://github.com/xbmc/xbmc/pull/20267 since it was reverted in https://github.com/xbmc/xbmc/pull/21308) (**edit:** Done in https://github.com/xbmc/xbmc/pull/21355)
- [x] Drop/deprecate the `Player.DisplayAfterSeek` boolean condition which is a current design flaw of VP (see background in https://github.com/xbmc/xbmc/pull/20753#issuecomment-1015707100). In this case the core is currently dictating for how long the GUI/Skin should display stuff when a seek happens. Affects for example EDL since the seekbar overlays the video for no reason. (Done in https://github.com/xbmc/xbmc/pull/21366 and https://github.com/xbmc/xbmc/pull/21380). Removal of displayafterseek from core done in https://github.com/xbmc/xbmc/pull/21425)

## Description

(c/p from the doxygen file)

Skin timers are skin objects that dependent on time and can be fully controlled from skins either using
builtin functions or boolean conditions/expressions. One can see them
as stopwatches that can be activated and deactivated automatically depending on the value of info expressions or simply activated/deactivated manually from builtins.

The framework was created to allow skins to control the visibility of windows (or controls) depending on
the elapsed time of timers the skin define. Skin timers allow multiple use cases in skins, previously only available via the execution of python scripts (or other hacks):
- Close a specific window after x seconds have elapsed
- Control the visibility of a group (or triggering an animation) depending on the elapsed time of a given timer
- Define a buffer time window that is kept activated for a short period of time (e.g. keep controls visible for x seconds after a player seek)
- Execute timed actions (on timer stop or timer start)
- etc

Skin timers are defined in a `Timers.xml` file within the xml directory of the skin. The file has the following "schema":

```xml
<timers>
    <timer>...</timer>
    <timer>...</timer>
</timers>
```

### Examples

The following example illustrates the simplest possible skin timer. This timer is completely manual (it has to be manually started and stopped):

```xml
<timer>
    <name>mymanualtimer</name>
    <description>100% manual timer</description>
</timer>
```

This timer can be controlled from your skin by executing the `Skin.TimerStart(mymanualtimer)` builtin or
the `Skin.TimerStop(mymanualtimer)` builtin. You can define the visibility of skin elements based on the internal
properties of the timer, such as the fact that the timer is active/running using `Skin.TimerIsRunning(mymanualtimer)` or depending on the elapsed time (e.g. 5 seconds) using the expression `Integer.IsGreaterOrEqual(Skin.TimerElapsedSecs(mymanualtimer),5)`.

The following timer is a variation of the previous timer with the particularity of being automatically stopped by the skinning engine after a maximum of elapsed 5 seconds without having to issue the `Skin.TimerStop(mymanualtimer)` builtin:

```xml
<timer>
    <name>mymanualautocloseabletimer</name>
    <description>100% manual autocloseable timer</description>
    <stop>Integer.IsGreaterOrEqual(Skin.TimerElapsedSecs(mymanualautocloseabletimer),5)</stop>
</timer>
```

This type of timer is particularly useful if you want to automatically close a specific window (or triggering a close animation) after x time has elapsed, while guaranteeing the timer is also stopped. See the example below which trigger a slide animation on a window after it has been open for 5 seconds:

```xml
<?xml version="1.0" encoding="utf-8"?>
<window type="dialog" id="1109">
	<onload>Skin.TimerStart(mymanualautocloseabletimer)</onload>
    ...
	<controls>
		<control type="group">
			<animation effect="slide" start="0,0" end="0,-80" time="300" condition="Integer.IsGreaterOrEqual(Skin.TimerElapsedSecs(mymanualautocloseabletimer),5)">Conditional</animation>
			...
        </control>
    </controls>
</window>
```

The following timer presents a notification (for 1 sec) whenever the timer is activated or deactivated:

```xml
<timer>
    <name>manualtimerwithactions</name>
    <description>100% manual timer with actions</description>
    <onstart>Notification(skintimer, My timer was started, 1000)</onstart>
    <onstop>Notification(skintimer, My timer was stopped, 1000)</onstop>
</timer>
```

The following timer is an example of a completely automatic timer. The timer is automatically activated or deactivated based on the value
of boolean info expressions. In this particular example, the timer is automatically started whenever the Player is playing a file (if not already running). It is stopped if
there is no file being played (and of course if previously running). Since the timer can be activated/deactivated multiple times, `reset="true"` ensures the timer is
always reset to 0 on each start operation. Whenever the timer is started or stopped, notifications are issued.

```xml
<timer>
    <name>myautomatictimer</name>
    <description>Player state checker</description>
    <start reset="true">Player.Playing</start>
    <stop>!Player.Playing</stop>
    <onstart>Notification(skintimer, Player has started playing a file, 1000)</onstart>
    <onstop>Notification(skintimer, Player is no longer playing a file, 1000)</onstop>
</timer>
```

In certain situations you might want to reset your timer without having to stop and start. For instance, if you want to stop the timer after 4 seconds
but have the timer resetting to 0 seconds if the user provides some input to Kodi. For such cases the `<reset/>` condition can be used:

```xml
<timer>
    <name>windowtimer</name>
    <description>Reset on idle</description>
    <start reset="true">Window.IsActive(mywindow)</start>
    <reset>Window.IsActive(mywindow) + !System.IdleTime(1) + Integer.IsGreaterOrEqual(Skin.TimerElapsedSecs(windowtimer), 1)</reset>
    <stop>!Window.IsActive(mywindow) + Integer.IsGreaterOrEqual(Skin.TimerElapsedSecs(windowtimer), 4)</stop>
    <onstop>Dialog.Close(mywindow)</onstop>
</timer>
```

### Available tags

| Tag           | Description                                                   |
|--------------:|:--------------------------------------------------------------|
| name          | The unique name of the timer. The name is used as the id of the timer, hence needs to be unique. <b>(required)</b>
| description   | The description of the timer, a helper string. <b>(optional)</b>
| start         | An info bool expression that the skinning engine should use to automatically start the timer <b>(optional)</b>
| reset         | An info bool expression that the skinning engine should use to automatically reset the timer <b>(optional)</b>
| stop          | An info bool expression that the skinning engine should use to automatically stop the timer <b>(optional)</b>
| onstart       | A builtin function that the skinning engine should execute when the timer is started <b>(optional)</b>
| onstop        | A builtin function that the skinning engine should execute when the timer is stopped <b>(optional)</b>

## How has this been tested?
Runtime tested with a skin timer implementation equivalent to https://github.com/xbmc/xbmc/pull/20816:

**Timers.xml**:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<timers>
    <timer>
        <name>topbaroverlaytimer</name>
        <description>Timer valid for 5 seconds while the Player is Playing</description>
        <stop>[!Player.Paused | Integer.IsGreaterOrEqual(Skin.TimerElapsedSecs(topbaroverlaytimer),5)]</stop>
        <onstart>Notification(skintimer, timer started, 1000)</onstart>
        <onstop>Notification(skintimer, timer stopped, 1000)</onstop>
    </timer>
</timers>
```
**Window:**

```xml
<?xml version="1.0" encoding="utf-8"?>
<window type="dialog" id="1109">
	<onload>Skin.TimerStart(topbaroverlaytimer)</onload>
	...
	<controls>
		<control type="group">
			<animation effect="slide" start="0,0" end="0,-80" time="300" condition="Player.Paused + Integer.IsGreaterOrEqual(Skin.TimerElapsedSecs(topbaroverlaytimer),5)">Conditional</animation>
                        ...
                </control>
	</controls>
</window>
```

I plan to implement the other 2 use-cases shortly. Since they will stress test the implementation and they may require changes to this one I suggest only to merge this once they are finished. Code review is appreciated though whenever you find time.

## What is the effect on users?
Skinners should now have a more powerful way of defining time-based logic in skins.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [x] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

File: xbmc/addons/gui/skin/SkinTimers.cpp [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+  m_stop = true;
+}
+
+void CSkinTimerManager::Process()
+{
+  while (!m_stop)
+  {
+    for (auto const& [key, val] : m_timers)
+    {
+      std::unique_ptr<CSkinTimer>::pointer timer = val.get();
```
Comment (ksooo) [inline xbmc/addons/gui/skin/SkinTimers.cpp:291]: const std::unique_ptr

File: xbmc/interfaces/builtins/SkinBuiltins.cpp [hunk 2]
```diff
@@ -458,6 +458,26 @@ static int SkinDebug(const std::vector<std::string>& params)
   return 0;
 }
 
+/*! \brief Starts a given skin timer
+ *  \param params The parameters.
+ *  \details params[0] = Name of the timer.
+ */
+static int SkinTimerStart(const std::vector<std::string>& params)
+{
+  g_SkinInfo->TimerStart(params[0]);
```
Comment (ksooo) [inline xbmc/interfaces/builtins/SkinBuiltins.cpp:467]: what, if params is empty?

File: xbmc/addons/gui/skin/SkinTimers.cpp [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+  Create();
+}
+
+void CSkinTimerManager::LoadTimers(const std::string& path)
+{
+  CXBMCTinyXML doc;
+  if (!doc.LoadFile(path))
+  {
+    CLog::LogF(LOGINFO, "Error loading timers file {}: {} (row: {}, col: {})", path,
```
Comment (ksooo) [inline xbmc/addons/gui/skin/SkinTimers.cpp:110]: LOGERROR, not INFO

File: xbmc/addons/gui/skin/SkinTimers.cpp [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+}
+
+void CSkinTimerManager::Start()
+{
+  if (m_timers.empty())
+  {
+    return;
+  }
+
+  Create();
```
Comment (ksooo) [inline xbmc/addons/gui/skin/SkinTimers.cpp:102]:
```c++
if (!m_timers.empty())
  Create();
```

... saves some lines of code

File: xbmc/addons/gui/skin/SkinTimers.cpp [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+  INFO::InfoPtr startInfo{nullptr};
+  INFO::InfoPtr stopInfo{nullptr};
+  std::string timerName = node->FirstChild("name")->FirstChild()->Value();
+  std::string startAction;
+  std::string stopAction;
+  bool reset{false};
+
+  if (m_timers.count(timerName) > 0)
+  {
+    CLog::LogF(LOGERROR,
```
Comment (ksooo) [inline xbmc/addons/gui/skin/SkinTimers.cpp:148]: LOGWARNING, as this seems not actually to be an error

File: xbmc/addons/gui/skin/SkinTimers.cpp [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+    {
+      stopAction = node->FirstChild("onstop")->FirstChild()->ValueStr();
+    }
+  }
+
+  m_timers[timerName] = std::make_unique<CSkinTimer>(
+      CSkinTimer(timerName, startInfo, stopInfo, startAction, stopAction, reset));
+}
+
+bool CSkinTimerManager::TimerIsRunning(const std::string& timer)
```
Comment (ksooo) [inline xbmc/addons/gui/skin/SkinTimers.cpp:207]: could this method be `const`?

File: xbmc/addons/gui/skin/SkinTimers.cpp [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+{
+  if (m_timers.count(timer) == 0)
+  {
+    CLog::LogF(LOGERROR, "Couldn't find Skin Timer with name: {}", timer);
+    return false;
+  }
+  return m_timers.at(timer)->IsRunning();
+}
+
+float CSkinTimerManager::GetTimerElapsedSeconds(const std::string& timer)
```
Comment (ksooo) [inline xbmc/addons/gui/skin/SkinTimers.cpp:217]: could this method be `const`?

File: xbmc/addons/gui/skin/SkinTimers.cpp [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+  }
+
+  // skintimers, as infomanager clients register info conditions/expressions in the infomanager.
+  // The infomanager is linked to skins, being initialized or cleared when
+  // skins are loaded (or unloaded). All the registered boolean conditions from
+  // skin timers will endup being removed when the skin is unloaded. However, to
+  // self-contain this component unregister them all here.
+  for (auto const& [key, val] : m_timers)
+  {
+    std::unique_ptr<CSkinTimer>::pointer timer = val.get();
```
Comment (ksooo) [inline xbmc/addons/gui/skin/SkinTimers.cpp:261]: const std::unique_ptr

File: xbmc/addons/gui/skin/SkinTimers.cpp (hunk 3 above)
Comment (enen92) [reply xbmc/addons/gui/skin/SkinTimers.cpp:110]: `LOGWARNING` probably better here as the file may not exist (and we removed the exist check)

File: xbmc/addons/gui/skin/SkinTimers.h [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+  */
+  CSkinTimer(const std::string& name,
+             const INFO::InfoPtr startCondition,
+             const INFO::InfoPtr stopCondition,
+             const std::string& startAction,
+             const std::string& stopAction,
+             bool resetOnStart);
+
+  /*! \brief Default skin timer destructor */
+  ~CSkinTimer() = default;
```
[the remark this answers is not in the document]
Comment (enen92) [reply xbmc/addons/gui/skin/SkinTimers.h:44]: `CStopWatch` (base class for `CSkinTimer`) destructor is not virtual we cannot override:

`error: ‘CSkinTimer::~CSkinTimer()’ marked ‘override’, but does not override`

Should I keep it as is?

File: xbmc/guilib/guiinfo/SkinGUIInfo.cpp [hunk 10]
```diff
@@ -111,6 +124,11 @@ bool CSkinGUIInfo::GetBool(bool& value, const CGUIListItem *gitem, int contextWi
       value = StringUtils::EqualsNoCase(theme, info.GetData3());
       return true;
     }
+    case SKIN_TIMER_IS_RUNNING:
+    {
+      value = g_SkinInfo->TimerIsRunning(info.GetData3());
```
[the remark this answers is not in the document]
Comment (enen92) [reply xbmc/guilib/guiinfo/SkinGUIInfo.cpp:129]: value is bool and `TimerIsRunning` returns a bool too, there is no implicit conversion here afaics

File: xbmc/interfaces/builtins/SkinBuiltins.cpp [hunk 11]
```diff
@@ -458,6 +458,26 @@ static int SkinDebug(const std::vector<std::string>& params)
...
+}
+
+/*! \brief Stops a given skin timer
+ *  \param params The parameters.
+ *  \details params[0] = Name of the timer.
+ */
+static int SkinTimerStop(const std::vector<std::string>& params)
+{
+  g_SkinInfo->TimerStop(params[0]);
+  return 0;
```
[the remark this answers is not in the document]
Comment (enen92) [reply xbmc/interfaces/builtins/SkinBuiltins.cpp:478]: I think this is by design unfortunately. Each builtin provider uses a command map to link to those static functions:

https://github.com/xbmc/xbmc/blob/a80d1293b0579bbed7dd7d20304ba34a20519f78/xbmc/interfaces/builtins/SystemBuiltins.cpp#L239-L257

CommandMap is `typedef std::map<std::string,CBuiltins::BUILT_IN> CommandMap;` while `CBuiltins::BUILT_IN` requires an (`int (*Execute)(const std::vector<std::string>& params);` function so those static methods have to return int.
It doesn't quite make sense, the only place where this return code is checked is one builtin:

```cpp
JSONRPC_STATUS CSystemOperations::EjectOpticalDrive(const std::string &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result)
{
  return CBuiltins::GetInstance().Execute("EjectTray") == 0 ? ACK : FailedToExecute;
}
```

Something to improve later but that I don't really want to change in this PR

File: xbmc/addons/gui/skin/SkinTimers.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/addons/gui/skin/SkinTimers.h:44]: Then, you should introduce a virtual ctor, should read: `virtual ~CSkinTimer() = default;`.

File: xbmc/guilib/guiinfo/SkinGUIInfo.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/guilib/guiinfo/SkinGUIInfo.cpp:129]: Ooops, my bad.

**Comment by jjd-uk** [comment]
Based on the Estuary PR's https://github.com/xbmc/xbmc/pull/21328 and https://github.com/xbmc/xbmc/pull/21355 the changes here seem to be working as expected. Can't review if the C++ changes are handled in the best way however.

**Comment by enen92** [comment]
All three use cases mentioned in the PR post have been implemented and validated the implementation.
Would be great to have reviews/approvals on this one so it can go in for alpha2

**Comment by basilgello** [comment]
Looks good to me from the brief look at the code. I will abstain from approval though until I make a Debian build.

File: xbmc/addons/gui/skin/SkinTimers.dox [hunk 12]
```diff
@@ -0,0 +1,145 @@
+/*!
```
Comment (enen92) [inline xbmc/addons/gui/skin/SkinTimers.dox:1]: @jjd-uk mind taking a look at the wording/English in this file please? Pretty sure I have typos here

Comment (enen92) [inline xbmc/addons/gui/skin/SkinTimers.dox:1]: @jjd-uk mind taking a look at the wording/English in this file please? Pretty sure I have typos here

File: xbmc/addons/gui/skin/SkinTimers.dox [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+    <name>myautomatictimer</name>
+    <description>Player state checker</description>
+    <start reset="true">Player.Playing</start>
+    <stop>!Player.Playing</stop>
+    <onstart>Notification(skintimer, Player has started playing a file, 1000)</onstart>
+    <onstop>Notification(skintimer, Player is no longer playing a file, 1000)</onstop>
+</timer>
+~~~~~~~~~~~~~
+
+In certain situations you might want to reset your timer without having to stop and start. For instance, if you want to stop the timer after 5 seconds
```
Comment (enen92) [inline xbmc/addons/gui/skin/SkinTimers.dox:110]: %s/5/4

Comment (enen92) [inline xbmc/addons/gui/skin/SkinTimers.dox:110]: %s/5/4

**Comment by jjd-uk** [comment]
@enen92 I'll try and find the time later to look at that doxygen file, if not later then tomorrow at latest.

FYI once this and your related Estuary PR's are in.

I've got this ready to PR

image

So ability to use or not use skin timer to autoclose seekbar when paused, and be able to configure the time after which autoclose takes effect.

I've also been able to work how to have a configurable DisplayAfterSeek replacement using skin timers, which is the **Seekbar autoclose after seeking time (seconds)**

**Comment by enen92** [comment]
Great! Too bad in the skin settings window we don't have access to setting levels (afaik). I am a bit afraid in that screenshots we clutter settings and add too much complexity.
No problem about time, take all you need.

**Comment by jjd-uk** [comment]
The additional complexity is why I'll leave it until your Estaury PR's are in, and then offer this as an enhancement. So nothing is held up if bike shedding starts on what I've done. If nothing else it will showcase to skinners the other possibilities such as doing a different displayafterseek replacement implementation.

File: xbmc/addons/gui/skin/SkinTimers.dox [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+~~~~~~~~~~~~~{.xml}
+<timer>
+    <name>mymanualautocloseabletimer</name>
+    <description>100% manual autocloseable timer</description>
+    <stop>Integer.IsGreaterOrEqual(Skin.TimerElapsedSecs(mymanualautocloseabletimer),5)</stop>
+</timer>
+~~~~~~~~~~~~~
+
+This type of timer is particularly usefull if you want to automatically close a specific window (or triggering a close animation) after x time has elapsed,
```
Comment (jjd-uk) [inline xbmc/addons/gui/skin/SkinTimers.dox:66]: useful

File: xbmc/addons/gui/skin/SkinTimers.dox [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+    <description>Player state checker</description>
+    <start reset="true">Player.Playing</start>
+    <stop>!Player.Playing</stop>
+    <onstart>Notification(skintimer, Player has started playing a file, 1000)</onstart>
+    <onstop>Notification(skintimer, Player is no longer playing a file, 1000)</onstop>
+</timer>
+~~~~~~~~~~~~~
+
+In certain situations you might want to reset your timer without having to stop and start. For instance, if you want to stop the timer after 5 seconds
+but have the timer reseting to 0 seconds if the user provides some input to Kodi. For such cases the `<reset/>` condition can be used:
```
Comment (jjd-uk) [inline xbmc/addons/gui/skin/SkinTimers.dox:111]: resetting

File: xbmc/addons/gui/skin/SkinTimers.dox [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+
+\page Skin_Timers Skin Timers
+\brief **Programatic time-based objects for Skins**
+
+\tableofcontents
+
+--------------------------------------------------------------------------------
+\section Skin_Timers_sect1 Description
+
+Skin timers are skin objects that dependent on time and can be fully controlled from skins either using
```
Comment (jjd-uk) [inline xbmc/addons/gui/skin/SkinTimers.dox:11]: that **are** dependent

File: xbmc/addons/gui/skin/SkinTimers.dox [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+\link page_List_of_built_in_functions **Builtin functions**\endlink or
+\link modules__infolabels_boolean_conditions **Infolabels and Boolean conditions**\endlink. One can see them
+as stopwatches that can be activated and deactivated automatically depending on the value of info expressions or simply activated/deactivated
+manually from builtins.
+The framework was created to allow skins to control the visibility of windows (and controls) depending on
+the elapsed time of timers the skin define. Skin timers allow multiple use cases in skins, previously only available via the execution
```
Comment (jjd-uk) [inline xbmc/addons/gui/skin/SkinTimers.dox:17]: skin **defines**.

File: xbmc/addons/gui/skin/SkinTimers.dox [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+of python scripts:
+- Close a specific window after x seconds have elapsed
```
Comment (jjd-uk) [inline xbmc/addons/gui/skin/SkinTimers.dox:19]: Closing

File: xbmc/addons/gui/skin/SkinTimers.dox [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+- Control the visibility of a group (or triggering an animation) depending on the elapsed time of a given timer
```
Comment (jjd-uk) [inline xbmc/addons/gui/skin/SkinTimers.dox:20]: Controlling

File: xbmc/addons/gui/skin/SkinTimers.dox [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+- Define a buffer time window that is kept activated for a short period of time (e.g. keep controls visible for x seconds after a player seek)
```
Comment (jjd-uk) [inline xbmc/addons/gui/skin/SkinTimers.dox:21]: Defining

File: xbmc/addons/gui/skin/SkinTimers.dox [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+- Execute timed actions (on timer stop or timer start)
```
Comment (jjd-uk) [inline xbmc/addons/gui/skin/SkinTimers.dox:22]: Executing

File: xbmc/addons/gui/skin/SkinTimers.dox [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+- etc
+
+Skin timers are defined in a `Timers.xml` file within the xml directory of the skin. The file has the following "schema":
```
Comment (jjd-uk) [inline xbmc/addons/gui/skin/SkinTimers.dox:25]: in **the** `Timers.xml`

File: xbmc/addons/gui/skin/SkinTimers.dox [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+    <description>100% manual timer</description>
+</timer>
+~~~~~~~~~~~~~
+
+This timer can be controlled from your skin by executing the \link Builtin_SkinStartTimer `Skin.TimerStart(mymanualtimer)` builtin\endlink or
+\link Builtin_SkinStopTimer `Skin.TimerStop(mymanualtimer)` builtin\endlink. You can define the visibility of skin elements based on the internal
+properties of the timer, such as the fact that the timer is active/running using \link Skin_TimerIsRunning `Skin.TimerIsRunning(mymanualtimer)` builtin\endlink
+or depending on the elapsed time (e.g. 5 seconds) using \link Skin_TimerElapsedSecs Integer.IsGreaterOrEqual(Skin.TimerElapsedSecs(mymanualtimer),5)\endlink.
+
+The following timer is a variation of the previous timer with the particularity of being automatically stopped by the skinning engine after a maximum of elapsed
```
Comment (jjd-uk) [inline xbmc/addons/gui/skin/SkinTimers.dox:55]: previous timer **but with the added ability of** being automatically stopped

File: xbmc/addons/gui/skin/SkinTimers.dox (hunk 12 above)
Comment (jjd-uk) [reply xbmc/addons/gui/skin/SkinTimers.dox:1]: Done, everything was perfectly understandable so just some minors.

Comment (enen92) [reply xbmc/addons/gui/skin/SkinTimers.dox:1]: Muchas gracias,
Updated all your comments :+1:

**Comment by enen92** [comment]
Merging by tomorrow night if no more feedback/requests

**Comment by enen92** [comment]
Today is yesterday's tomorrow night :)
Merging.

**Pull request merged**